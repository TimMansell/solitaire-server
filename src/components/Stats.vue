<template>
  <div class="stats">
    <div class="stats__games">
      Games played: <span data-test="stats">{{ userStats.gameNumber }}</span>
    </div>
    <div>
      Time:
      <span data-test="timer">{{ timer }}s</span>
    </div>
  </div>
</template>

<script>
import { mapGetters } from 'vuex';

export default {
  name: 'Stats',
  computed: {
    ...mapGetters(['userStats', 'timer']),
  },
};
</script>

<style scoped lang="scss">
.stats {
  display: flex;
  margin-bottom: var(--mg-sm);
  color: var(--text-primary);
  font-size: var(--font-size);

  @media (min-width: $bp-md) {
    display: block;
    margin-bottom: var(--mg-md);
    font-size: var(--font-size-lg);
  }

  &__games {
    flex: 1;
  }
}
</style>
