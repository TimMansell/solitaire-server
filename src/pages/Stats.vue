<template>
  <GameOverlay center-content show-close data-test="stats-overlay">
    <template #title> Your Stats </template>
    <template #msg>
      <Table
        :headings="['Played', 'Won', 'Lost', 'Gave Up']"
        :items="userStatstics"
        :placeholder-rows="2"
        data-test="user-stats"
      />
      <h1>Global Stats</h1>
      <Table
        :headings="['Played', 'Won', 'Lost', 'Gave Up']"
        :items="globalStatstics"
        :placeholder-rows="2"
        data-test="global-stats"
      />
    </template>
  </GameOverlay>
</template>

<script>
import numeral from 'numeral';
import { mapActions, mapGetters } from 'vuex';
import GameOverlay from '@/components/GameOverlay.vue';
import Table from '@/components/Table.vue';

export const calcPercent = (value) => numeral(value).format('0.00%');

export const calcNumber = (value) => numeral(value).format('0,0');

export const calcStats = ({ completed, won, lost }) => {
  if (!completed && completed !== 0) {
    return [];
  }

  const abandoned = completed - won - lost;

  const completedCount = calcNumber(completed);
  const wonCount = calcNumber(won);
  const lostCount = calcNumber(lost);
  const abandonedCount = calcNumber(abandoned);

  const wonPercent = calcPercent(won / completed);
  const lostPercent = calcPercent(lost / completed);
  const abandonedPercent = calcPercent(abandoned / completed);

  const stats = [
    [completedCount, wonCount, lostCount, abandonedCount],
    ['', wonPercent, lostPercent, abandonedPercent],
  ];

  return stats;
};

export default {
  name: 'GameLost',
  components: {
    GameOverlay,
    Table,
  },
  computed: {
    ...mapGetters(['userStats', 'globalStats']),
    userStatstics() {
      const { userStats } = this;
      const stats = calcStats(userStats);

      return stats;
    },
    globalStatstics() {
      const { globalStats } = this;
      const stats = calcStats(globalStats);

      return stats;
    },
  },
  methods: {
    ...mapActions(['getStats', 'clearStats']),
  },
  mounted() {
    this.getStats();
  },
  destroyed() {
    this.clearStats();
  },
};
</script>
