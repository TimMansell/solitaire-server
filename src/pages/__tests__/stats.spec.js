import { shallowMount } from '@vue/test-utils';
import Stats from '@/pages/Stats.vue';

describe('Stats.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(Stats);

    expect(wrapper).toMatchSnapshot();
  });
});
