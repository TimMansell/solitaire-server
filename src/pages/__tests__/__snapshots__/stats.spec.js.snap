// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats.vue matches snapshot 1`] = `
<gameoverlay-stub centercontent="true" showclose="true" data-test="stats-overlay"> Your Stats <stats-stub></stats-stub>
</gameoverlay-stub>
`;
